flink-runtime/src/main/java/org/apache/flink/runtime/iterative/task/AbstractIterativeTask.java

```diff
@@ -183,8 +186,9 @@ protected void closeLocalStrategiesAndCaches() {
 	@Override
 	public DistributedRuntimeUDFContext createRuntimeContext(MetricGroup metrics) {
 		Environment env = getEnvironment();
+
 		return new IterativeRuntimeUdfContext(env.getTaskInfo(), getUserCodeClassLoader(),
-				getExecutionConfig(), env.getDistributedCacheEntries(), this.accumulatorMap, metrics);
+				getExecutionConfig(), env.getDistributedCacheEntries(), this.accumulatorMap, metrics, ExternalResourceUtils.getExternalResourceInfo(env.getExternalResourceDrivers(), env.getTaskManagerInfo().getConfiguration()));
```

Comment:
How expensive do we expect calls to the `ExternalResourceDrivers` to be? I'm just wondering because here we retrieve the `ExternalResourceInfos` for every task, independent of whether they need it or not. I think I would prefer that this feature would not add additional costs for tasks which don't want to use it.

Reply:
I think it would not be expensive since we only retrieve the `ExternalResourceInfos` at the first time. After that, it will use the cache in `SharedExternalResources`.

Reply:
But we don't know what the `ExternalResourceDriver` implementation does. This can be arbitrary complex. I think we should be super careful with introducing additional computational actions. Moreover, they should only take effect if the user really needs them.